Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php

```diff
@@ -92,23 +92,41 @@ class ContentCacheFlusher
      * model or the Neos Publishing Service.
      *
      * @param NodeInterface $node The node which has changed in some way
+     * @param Workspace $targetWorkspace An optional workspace to flush
      * @return void
      * @throws \Neos\ContentRepository\Exception\NodeTypeNotFoundException
      */
-    public function registerNodeChange(NodeInterface $node)
+    public function registerNodeChange(NodeInterface $node, Workspace $targetWorkspace = null)
     {
         $this->tagsToFlush[ContentCache::TAG_EVERYTHING] = 'which were tagged with "Everything".';
 
         if (empty($this->workspacesToFlush[$node->getWorkspace()->getName()])) {
             $this->resolveWorkspaceChain($node->getWorkspace());
         }
 
+        if ($targetWorkspace !== null && empty($this->workspacesToFlush[$targetWorkspace->getName()])) {
+            $this->resolveWorkspaceChain($targetWorkspace);
+        }
+
         if (!array_key_exists($node->getWorkspace()->getName(), $this->workspacesToFlush)) {
             return;
         }
 
+        $this->registerAllTagsToFlushForNodeInWorkspace($node, $node->getWorkspace());
+        if ($targetWorkspace !== null) {
+            $this->registerAllTagsToFlushForNodeInWorkspace($node, $targetWorkspace);
+        }
+    }
+
+    /**
+     * @param NodeInterface $node
+     * @param Workspace $workspace
+     */
+    protected function registerAllTagsToFlushForNodeInWorkspace(NodeInterface $node, Workspace $workspace)
+    {
         $nodeIdentifier = $node->getIdentifier();
-        foreach ($this->workspacesToFlush[$node->getWorkspace()->getName()] as $workspaceName => $workspaceHash) {
+
+        foreach ($this->workspacesToFlush[$workspace->getName()] as $workspaceName => $workspaceHash) {
```

Comment:
Should we check if key exists / is array before itterating over it?

Reply:
Why does `$this->workspacesToFlush` contain lists of workspaces for a workspace? I'm puzzled… anyways:

```suggestion
        if (!array_key_exists($workspace->getName(), $this->workspacesToFlush) || is_array($this->workspacesToFlush[$workspace->getName()]) === false) {
            return;
        }

        foreach ($this->workspacesToFlush[$workspace->getName()] as $workspaceName => $workspaceHash) {
```